src/main/java/com/qiniu/storage/ResumeUploadPerformer.java

```diff
@@ -0,0 +1,210 @@
+package com.qiniu.storage;
+
+import com.google.gson.Gson;
+import com.qiniu.common.Constants;
+import com.qiniu.common.QiniuException;
+import com.qiniu.http.Client;
+import com.qiniu.http.Response;
+import com.qiniu.util.StringMap;
+import com.qiniu.util.StringUtils;
+
+import java.io.IOException;
+
+abstract class ResumeUploadPerformer {
+
+    private final Client client;
+    private final Recorder recorder;
+    private final Configuration config;
+    private final ConfigHelper configHelper;
+
+    final String key;
+    final UploadToken token;
+    final ResumeUploadSource uploadSource;
+    final UploadOptions options;
+
+    ResumeUploadPerformer(Client client, String key, UploadToken token, ResumeUploadSource source,
+                          Recorder recorder, UploadOptions options, Configuration config) {
+        this.client = client;
+        this.key = key;
+        this.token = token;
+        this.uploadSource = source;
+        this.options = options == null ? UploadOptions.defaultOptions() : options;
+        this.recorder = recorder;
+        this.config = config;
+        this.configHelper = new ConfigHelper(config);
+    }
+
+    boolean isAllBlocksUploadingOrUploaded() {
+        return uploadSource.isAllBlocksUploadingOrUploaded();
+    }
+
+    boolean isAllBlocksUploaded() {
+        return uploadSource.isAllBlocksUploaded();
+    }
+
+
+    abstract boolean shouldInit();
+
+    abstract Response uploadInit() throws QiniuException;
+
+    Response uploadNextData() throws QiniuException {
+        ResumeUploadSource.Block block = null;
+        synchronized (this) {
+            block = getNextUploadingBlock();
+            if (block != null) {
+                block.isUploading = true;
+            }
+        }
+
+        if (block == null) {
+            return Response.createSuccessResponse();
+        }
+
+        System.out.printf("上传块开始：index:%d offset:%d size:%d \r\n", block.index, block.offset, block.size);
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.